### PR TITLE
Replace Contract.Assert with Debug.Assert

### LINKED ISSUE #682: Contract.Assert vs Debug.Assert
Some of the CoreFX libraries use Contract.Assert.
Others use Debug.Assert.
We should be consistent, and if possible sweep the source to unify this.

### PR DESCRIPTION
**Pull request by ellismg**
Fixes #682

### CODE REVIEW

**Comment by ellismg** [comment]
Let me know if you'd rather have this against master.  Should be easy enough to rebase into there.  I spoke with @jaredpar about doing this in the CodeFormatter (my feeling was the changes here are semantic in nature and it seemed out of the purview of the tool) and he agreed.  A little git magic got us most of the way there and a little emacs magic finished off the job.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/PartitionerStatic.cs [hunk 1]
```diff
@@ -1160,13 +1160,13 @@ override protected bool HasNoElementsLeft
...
                 {
-                    Contract.Assert(_sharedIndex != null);
+                    Debug.Assert(_sharedIndex != null);
                     // use the new Volatile.Read method because it is cheaper than Interlocked.Read on AMD64 architecture
                     return Volatile.Read(ref _sharedIndex.Value) >= SourceCount - 1;
                 }
                 set
                 {
-                    Contract.Assert(false);
+                    Debug.Assert(false);
```
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/PartitionerStatic.cs]: There are ~30 cases of Debug.Assert(false, and I believe all but this one (at least in the files you modified) already have a message argument.  Could we change all of them to use Debug.Fail instead?

File: src/System.Numerics.Vectors/src/System/Numerics/Register.tt [hunk 2]
```diff
@@ -5,7 +5,7 @@
 <#@ import namespace="System" #>
 <#@ import namespace="System.Linq" #>
 <#@ import namespace="System.Runtime.InteropServices" #>
-<#@ import namespace="System.Diagnostics.Contracts" #>
```
Comment (stephentoub) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: Are there no .cs files that need to be regenerated based on these template changes?  And if not, does that mean Register.tt is dead code?

File: src/System.Linq.Parallel/src/System/Linq/Parallel/Utils/Sorting.cs [hunk 3]
```diff
@@ -59,18 +59,18 @@ internal class SortHelper<TInputOutput, TKey> : SortHelper<TInputOutput>, IDispo
...
+            Debug.Assert(sharedIndices != null);
+            Debug.Assert(sharedkeys != null);
+            Debug.Assert(sharedValues != null);
+            Debug.Assert(sharedBarriers != null);
+            Debug.Assert(sharedIndices.Length <= sharedkeys.Length);
+            Debug.Assert(sharedIndices.Length == sharedValues.Length);
             // Multi-dim arrays are simulated using jagged arrays.
             // Because of that, when phaseCount == 0, we end up with an empty sharedBarrier array.
             // Since there are no cases when phaseCount == 0 where we actually access the sharedBarriers, I am removing this check.
-            // Contract.Assert(sharedIndices.Length == sharedBarriers[0].Length);
```
Comment (stephentoub) [inline src/System.Linq.Parallel/src/System/Linq/Parallel/Utils/Sorting.cs]: There are a handful of commented out asserts in these files.  Should we just delete them as part of this change?

**Comment by stephentoub** [comment]
A few minor comments, but otherwise LGTM.  Thanks for doing this, Matt.

File: src/System.Numerics.Vectors/src/System/Numerics/Register.tt [hunk 4]
```diff
@@ -5,7 +5,7 @@
...
+<#@ import namespace="System.Diagnostics" #>
```
Comment (crowwhite) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: wan hussin abdul rahman

File: src/System.Numerics.Vectors/src/System/Numerics/Register.tt (hunk 2 above)
Comment (nguerrera) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: cc @mellinoe

Comment (Raverenx) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: Register.cs needs to be regenerated with this tt file, it's not dead code.

Comment (mellinoe) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: Register.cs needs to be regenerated along with this tt file; it's not dead code.

Comment (ellismg) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: I believe that the Assert in this file is something checked at template instantiation time not runtime.  Contract.cs doesn't have any Debug or Contract asserts in it and then test refers to variables introduced by GenerationConfiguration.ttinclude.

File: src/System.Linq.Parallel/src/System/Linq/Parallel/Utils/Sorting.cs (hunk 3 above)
Comment (ellismg) [inline src/System.Linq.Parallel/src/System/Linq/Parallel/Utils/Sorting.cs]: I'm not going to do this as part of the change.

I looked at the places we had "// Debug.Assert" and there were two hits (obviously there could be other cases that miss such a simple search).  In both cases it seemed reasonable to keep the assert commented out (here there's an explanation as to why it is not enabled, the other was an assert that we would need to enable if we extend the code).  I would be fine if the owners of the code want to remove them, but it felt a little weird to do as part of this larger (mostly mechanical) change.

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/PartitionerStatic.cs (hunk 1 above)
Comment (ellismg) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/PartitionerStatic.cs]: Done.

File: src/System.Numerics.Vectors/src/System/Numerics/Register.tt (hunk 2 above)
Comment (mellinoe) [inline src/System.Numerics.Vectors/src/System/Numerics/Register.tt]: Yes, you're right about that; I didn't look closely enough here. It's part of the template code so it won't have an effect on any of the generated code.

**Comment by nguerrera** [comment]
As discussed offline, please resubmit to master. I'm deleting v1.0 now.

**Pull request closed without merging**